static/js/lib/coupon.js

```diff
@@ -0,0 +1,91 @@
+// @flow
+import {
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_CONTENT_TYPE_COURSE,
+  COUPON_CONTENT_TYPE_COURSERUN,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../constants';
+import type {
+  Coupons,
+  Coupon,
+  CalculatedPrices,
+} from '../flow/couponTypes';
+import type {
+  CoursePrice,
+  CoursePrices,
+  Dashboard,
+} from '../flow/dashboardTypes';
+
+export const calculatePrices = (programs: Dashboard, prices: CoursePrices, coupons: Coupons): CalculatedPrices => {
+  type HasProgramId = {
+    program_id: number
+  };
+  function makeLookup<T: HasProgramId>(obj: Array<T>): Map<number, T> {
+    return new Map(obj.map((value: T) => [value.program_id, value]));
+  }
+  let couponLookup: Map<number, Coupon> = makeLookup(coupons);
+  let priceLookup: Map<number, CoursePrice> = makeLookup(prices);
+
+  let calcPriceForRun = (run, course, program) => ({
+    id: run.id,
+    price: mockableCalculateRunPrice(
+      run.id, course.id, program.id, priceLookup.get(program.id), couponLookup.get(program.id)
+    ),
+  });
+
+  let calcPriceForCourse = (course, program) => ({
+    id: course.id,
+    runs: course.runs.map(run => calcPriceForRun(run, course, program))
+  });
+
+  let calcPriceForProgram = program => ({
+    id: program.id,
+    courses: program.courses.map(course => calcPriceForCourse(course, program))
+  });
```

Comment:
I think for writing these the ramda functions `over` and `lensProp` might be really helpful, `over` basically lets you 'edit' an object by supplying an update func and a lens that focuses on the property that you want to update. The return value will be a clone of the object, with the value of the focused prop set to the value of the update function called with that prop's original value as an argument.

You could do something like:

```js
let calcPriceforProgram = program => (
  R.over(R.lensProp('courses'), R.map(calcPriceForCourse(program)), program)
);
```

(assuming the argument order for `calcPriceForCourse` could be switched and updated to be a curried function). You could use a similar pattern for the `calcPriceForCourse` and `calcPriceForRun` function as well.